File: hub/adapters/stt/deepgram_live.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
import json
from time import perf_counter
from typing import Any

import websockets
from websockets.exceptions import ConnectionClosed
from websockets.protocol import State

from hub.adapters.interfaces import TranscriptResult


@dataclass(slots=True)
class _TurnState:
    conversation_id: str
    started_at: float = field(default_factory=perf_counter)
    final_segments: list[str] = field(default_factory=list)
    interim_text: str = ""
    finalize_requested: bool = False
    finalize_event: asyncio.Event = field(default_factory=asyncio.Event)


class DeepgramLiveSTTProvider:
    def __init__(
        self,
        *,
        api_key: str,
        model: str = "nova-3",
        sample_rate: int = 16000,
        endpointing_ms: int = 300,
        utterance_end_ms: int = 1000,
    ) -> None:
        self._api_key = api_key
        self._model = model
        self._sample_rate = sample_rate
        self._endpointing_ms = endpointing_ms
        self._utterance_end_ms = utterance_end_ms
        self._ws: websockets.ClientConnection | None = None
        self._reader_task: asyncio.Task[None] | None = None
        self._keepalive_task: asyncio.Task[None] | None = None
        self._active_turn: _TurnState | None = None
        self._send_lock = asyncio.Lock()
        self._last_send_at = perf_counter()

    async def start_turn(self, conversation_id: str) -> None:
        if self._active_turn is not None:
            raise RuntimeError("Deepgram turn already active")
        await self._ensure_connected()
        self._active_turn = _TurnState(conversation_id=conversation_id)

    async def send_audio(self, data: bytes) -> None:
        if not data or self._active_turn is None:
            return
        await self._send_binary(data)

    async def finish_turn(self) -> TranscriptResult:
        turn = self._active_turn
        if turn is None:
            return TranscriptResult(text="", provider="deepgram-live", latency_ms=0, is_final=True)

        turn.finalize_requested = True
        try:
            await self._send_text({"type": "Finalize"})
            await asyncio.wait_for(turn.finalize_event.wait(), timeout=2.5)
        except (TimeoutError, ConnectionClosed):
            pass
        finally:
            self._active_turn = None

        transcript = " ".join(segment for segment in turn.final_segments if segment).strip()
        if not transcript:
            transcript = turn.interim_text.strip()
        return TranscriptResult(
            text=transcript,
            provider="deepgram-live",
            latency_ms=int((perf_counter() - turn.started_at) * 1000),
            is_final=True,
        )

    async def aclose(self) -> None:
        if self._keepalive_task is not None:
            self._keepalive_task.cancel()
            self._keepalive_task = None
        if self._reader_task is not None:
            self._reader_task.cancel()
            self._reader_task = None
        if self._ws is not None:
            await self._ws.close()
            self._ws = None
        self._active_turn = None

    async def _ensure_connected(self) -> None:
        if _is_open(self._ws):
            return
        url = (
            "wss://api.deepgram.com/v1/listen"
            f"?model={self._model}"
            "&encoding=linear16"
            f"&sample_rate={self._sample_rate}"
            "&channels=1"
            "&interim_results=true"
            f"&endpointing={self._endpointing_ms}"
            f"&utterance_end_ms={self._utterance_end_ms}"
            "&vad_events=true"
            "&smart_format=true"
        )
        self._ws = await websockets.connect(
            url,
            additional_headers={"Authorization": f"Token {self._api_key}"},
            ping_interval=20,
            ping_timeout=20,
            max_size=None,
        )
        self._reader_task = asyncio.create_task(self._reader_loop())
        self._keepalive_task = asyncio.create_task(self._keepalive_loop())

    async def _reader_loop(self) -> None:
        try:
            assert self._ws is not None
            async for raw_message in self._ws:
                if isinstance(raw_message, bytes):
                    continue
                self._handle_message(json.loads(raw_message))
        except asyncio.CancelledError:
            return
        except ConnectionClosed:
            return
        finally:
            self._ws = None

    def _handle_message(self, payload: dict[str, Any]) -> None:
        turn = self._active_turn
        if turn is None:
            return

        message_type = payload.get("type", "")
        if message_type == "UtteranceEnd":
            if turn.finalize_requested:
                turn.finalize_event.set()
            return

        if message_type != "Results":
            return

        channel = payload.get("channel", {})
        alternatives = channel.get("alternatives", [])
        if not alternatives:
            return

        transcript = str(alternatives[0].get("transcript", "") or "").strip()
        if transcript:
            if payload.get("is_final"):
                if not turn.final_segments or turn.final_segments[-1] != transcript:
                    turn.final_segments.append(transcript)
            else:
                turn.interim_text = transcript

        if turn.finalize_requested and (
            payload.get("speech_final") or payload.get("from_finalize")
        ):
            turn.finalize_event.set()

    async def _send_binary(self, data: bytes) -> None:
        async with self._send_lock:
            await self._ensure_connected()
            assert self._ws is not None
            await self._ws.send(data)
            self._last_send_at = perf_counter()

    async def _send_text(self, payload: dict[str, Any]) -> None:
        async with self._send_lock:
            await self._ensure_connected()
            assert self._ws is not None
            await self._ws.send(json.dumps(payload))
            self._last_send_at = perf_counter()

    async def _keepalive_loop(self) -> None:
        try:
            while True:
                await asyncio.sleep(8)
                if not _is_open(self._ws):
                    return
                if perf_counter() - self._last_send_at < 8:
                    continue
                try:
                    await self._send_text({"type": "KeepAlive"})
                except ConnectionClosed:
                    return
        except asyncio.CancelledError:
            return


def _is_open(connection: websockets.ClientConnection | None) -> bool:
    return connection is not None and getattr(connection, "state", None) == State.OPEN

File: src/ts/hub/deepgram-live.ts
import WebSocket from "ws";

export interface TranscriptInterim {
  text: string;
  provider: string;
  latencyMs: number;
}

export interface TranscriptResult {
  text: string;
  provider: string;
  latencyMs: number;
}

interface DeepgramRealtimeCallbacks {
  onInterim?: (event: TranscriptInterim) => void;
  onUtterance?: (event: TranscriptResult) => void;
  onError?: (error: Error) => void;
}

export class DeepgramRealtimeSession {
  private ws: WebSocket | null = null;
  private keepAliveTimer: NodeJS.Timeout | null = null;
  private lastSendAt = Date.now();
  private readonly finalSegments: string[] = [];
  private interimText = "";
  private utteranceStartedAt = 0;
  private readonly endpointingMs = parsePositiveInt(process.env.DEEPGRAM_ENDPOINTING_MS, 800);
  private readonly utteranceEndMs = parsePositiveInt(process.env.DEEPGRAM_UTTERANCE_END_MS, 1800);

  constructor(
    private readonly apiKey: string,
    private readonly sampleRate: number,
    private readonly callbacks: DeepgramRealtimeCallbacks,
    private readonly model = "nova-3",
  ) {}

  async start(): Promise<void> {
    const url =
      "wss://api.deepgram.com/v1/listen" +
      `?model=${encodeURIComponent(this.model)}` +
      "&encoding=linear16" +
      `&sample_rate=${this.sampleRate}` +
      "&channels=1" +
      "&interim_results=true" +
      `&endpointing=${this.endpointingMs}` +
      `&utterance_end_ms=${this.utteranceEndMs}` +
      "&vad_events=true" +
      "&smart_format=true";

    this.ws = new WebSocket(url, {
      headers: {
        Authorization: `Token ${this.apiKey}`,
      },
    });
    this.ws.on("message", (data) => {
      try {
        this.handleMessage(String(data));
      } catch (error) {
        this.callbacks.onError?.(error instanceof Error ? error : new Error(String(error)));
      }
    });
    this.ws.on("error", (error) => {
      this.callbacks.onError?.(error instanceof Error ? error : new Error(String(error)));
    });
    await waitForOpen(this.ws);
    this.keepAliveTimer = setInterval(() => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        return;
      }
      if ((Date.now() - this.lastSendAt) < 8000) {
        return;
      }
      this.ws.send(JSON.stringify({ type: "KeepAlive" }));
    }, 8000);
  }

  sendAudio(chunk: Buffer): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN || chunk.length === 0) {
      return;
    }
    if (this.utteranceStartedAt === 0) {
      this.utteranceStartedAt = Date.now();
    }
    this.ws.send(chunk);
    this.lastSendAt = Date.now();
  }

  async close(): Promise<void> {
    if (this.keepAliveTimer) {
      clearInterval(this.keepAliveTimer);
      this.keepAliveTimer = null;
    }
    const ws = this.ws;
    this.ws = null;
    if (!ws) {
      return;
    }
    await new Promise<void>((resolve) => {
      if (ws.readyState === WebSocket.CLOSED) {
        resolve();
        return;
      }
      ws.once("close", () => resolve());
      ws.close();
    });
  }

  private handleMessage(raw: string): void {
    const payload = JSON.parse(raw) as Record<string, unknown>;
    const type = String(payload.type || "");
    if (type === "UtteranceEnd") {
      this.flushUtterance();
      return;
    }
    if (type !== "Results") {
      return;
    }

    const channel = (payload.channel || {}) as Record<string, unknown>;
    const alternatives = (channel.alternatives || []) as Array<Record<string, unknown>>;
    const transcript = String(alternatives[0]?.transcript || "").trim();
    const isFinal = Boolean(payload.is_final);
    const speechFinal = Boolean(payload.speech_final || payload.from_finalize);

    if (transcript) {
      if (this.utteranceStartedAt === 0) {
        this.utteranceStartedAt = Date.now();
      }
      if (isFinal) {
        if (this.finalSegments[this.finalSegments.length - 1] !== transcript) {
          this.finalSegments.push(transcript);
        }
        this.interimText = "";
      } else {
        this.interimText = transcript;
        const combined = [...this.finalSegments, transcript].join(" ").trim();
        if (combined) {
          this.callbacks.onInterim?.({
            text: combined,
            provider: "deepgram-live",
            latencyMs: this.utteranceLatencyMs(),
          });
        }
      }
    }

    if (speechFinal) {
      this.flushUtterance();
    }
  }

  private flushUtterance(): void {
    const text = [...this.finalSegments, this.finalSegments.length === 0 ? this.interimText : ""]
      .join(" ")
      .trim();
    if (!text) {
      this.resetUtterance();
      return;
    }
    this.callbacks.onUtterance?.({
      text,
      provider: "deepgram-live",
      latencyMs: this.utteranceLatencyMs(),
    });
    this.resetUtterance();
  }

  private utteranceLatencyMs(): number {
    return this.utteranceStartedAt > 0 ? Date.now() - this.utteranceStartedAt : 0;
  }

  private resetUtterance(): void {
    this.finalSegments.length = 0;
    this.interimText = "";
    this.utteranceStartedAt = 0;
  }
}

function parsePositiveInt(raw: string | undefined, fallback: number): number {
  const value = Number.parseInt(raw || "", 10);
  return Number.isInteger(value) && value > 0 ? value : fallback;
}

async function waitForOpen(ws: WebSocket): Promise<void> {
  if (ws.readyState === WebSocket.OPEN) {
    return;
  }
  await new Promise<void>((resolve, reject) => {
    ws.once("open", () => resolve());
    ws.once("error", (error) => reject(error));
  });
}
